src/github.com/matrix-org/dendrite/roomserver/storage/storage.go

```diff
@@ -195,3 +195,62 @@ func (d *Database) StateBlockNIDs(stateNIDs []types.StateSnapshotNID) ([]types.S
 func (d *Database) StateEntries(stateBlockNIDs []types.StateBlockNID) ([]types.StateEntryList, error) {
 	return d.statements.bulkSelectStateDataEntries(stateBlockNIDs)
 }
+
+// GetLatestEventsForUpdate implements input.EventDatabase
+func (d *Database) GetLatestEventsForUpdate(roomNID types.RoomNID) ([]types.StateAtEventAndReference, types.RoomRecentEventsUpdater, error) {
+	txn, err := d.db.Begin()
+	if err != nil {
+		return nil, nil, err
+	}
+	eventNIDs, err := d.statements.selectLatestEventsNIDsForUpdate(txn, roomNID)
+	if err != nil {
+		txn.Rollback()
+		return nil, nil, err
+	}
+	stateAndRefs, err := d.statements.bulkSelectStateAtEventAndReference(txn, eventNIDs)
+	if err != nil {
+		txn.Rollback()
```

Comment:
We need to end the txn somehow, and Rollback is probably clearer than Commit here?